site/docs/samples/notebook.md

```diff
@@ -1,162 +1,241 @@
+      ```
+      ACCESS_KEY=$(kubectl get secret minio -o jsonpath="{.data.accesskey}" | base64 --decode)
+      SECRET_KEY=$(kubectl get secret minio -o jsonpath="{.data.secretkey}" | base64 --decode)
+      echo "Access Key: ${ACCESS_KEY}"
+      echo "Secret Key: ${SECRET_KEY}"
+      ```
+    1. Click the :fontawesome-solid-plus-circle: button in the bottom right corner and then **Create bucket** to create a bucket (e.g. "demo").
+    1. Click the :fontawesome-solid-plus-circle: button again and then **Upload files** to upload the CSV file to the newly created bucket.
     
+## Register the dataset in a data catalog
+
+Register the credentials required for accessing the dataset. Replace the values for `access_key` and `secret_key` with the values from the object storage service that you used and run:
+
+```yaml
+cat << EOF | kubectl apply -f -
+apiVersion: v1
+kind: Secret
+metadata:
+  name: data-csv
+type: Opaque
+stringData:
+  access_key: "${ACCESS_KEY}"
+  secret_key: "${SECRET_KEY}"
+EOF
+```
+
+Then, register the data asset itself in the catalog. Replace the values for `endpoint`, `bucket` and `objectKey` with values from the object storage service that you used and run:
+
+```yaml
+cat << EOF | kubectl apply -f -
+apiVersion: katalog.m4d.ibm.com/v1alpha1
+kind: Asset
+metadata:
+  name: data-csv
+spec:
+  secretRef: 
+    name: data-csv
+  assetDetails:
+    dataFormat: csv
+    connection:
+      type: s3
+      s3:
+        endpoint: "http://minio.m4d-notebook-sample.svc.cluster.local:9000"
+        bucket: "demo"
+        objectKey: "data.csv"
+  assetMetadata:
+    geography: us-south
+    tags:
+    - finance
+    componentsMetadata:
+      nameOrig: 
+        tags:
+        - PII
+      oldbalanceOrg:
+        tags:
+        - sensitive
+      newbalanceOrig:
+        tags:
+        - sensitive
+EOF
+```
+
+The asset is now registered in the catalog. The identifier of the asset is `m4d-notebook-sample/data-csv` (i.e. `<namespace>/<name>`). You will use that name in the `M4DApplication` later.
+
+Notice the `assetMetadata` field above. It specifies the dataset geography and tags. These attributes can later be used in policies.
+
+
+## Define data access policies
+
+Define an [OpenPolicyAgent](https://www.openpolicyagent.org/) policy to redact the columns `nameOrig` and `nameDest` for datasets tagged as `finance`. Below is the policy (written in [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/#what-is-rego) language):
+
+```rego
+package dataapi.authz
+
+import data.data_policies as dp
+
+transform[action] {
+  description = "Redact sensitive columns in finance datasets"
+
+  dp.check_access_type([dp.AccessTypes.READ])
+  dp.dataset_has_tag("finance")
+  column_names := dp.column_with_any_name({"nameOrig", "nameDest"})
+
+  action = dp.build_redact_column_action(column_names[_], dp.build_policy_from_description(description))
+}
+```
```

Comment:
@rohithdv can I get you help to understand why this policy doesn't take affect? I tried without "nameDest" too. The confimap status indicates that it was loaded by opa successfully. However, requests still return allow. 

Input extracted from opa connector logs:

```json
{ "input": {
        "dataset_id": "m4d-notebook-sample/data-csv",
        "destination": "theshire",
        "details": {
                "credentials_info": {
                        "vault_secret_path": "/v1/kubernetes-secrets/data-csv?namespace=m4d-notebook-sample"
                },
                "data_format": "csv",
                "data_store": {
                        "name": "data-csv",
                        "s3": {
                                "bucket": "notebook-sample",
                                "endpoint": "s3.eu-gb.cloud-object-storage.appdomain.cloud",
                                "object_key": "data.csv"
                        },
                        "type": 2
                },
                "geo": "us-south",
                "metadata": {
                        "components_metadata": {
                                "nameOrig": {
                                        "component_type": "column",
                                        "tags": [
                                                "PII"
                                        ]
                                },
                                "newbalanceOrig": {
                                        "component_type": "column",
                                        "tags": [
                                                "sensitive"
                                        ]
                                },
                                "oldbalanceOrg": {
                                        "component_type": "column",
                                        "tags": [
                                                "sensitive"
                                        ]
                                }
                        },
                        "dataset_tags": [
                                "finance"
                        ]                        
                },
                "name": "m4d-notebook-sample/data-csv"
        },
        "processing_geography": "theshire",
        "properties": {
                "intent": "fraud-detection"
        },
        "type": "READ"
} }
```

Reply:
I see that there is no Role field. I am not sure we handle those cases now. Can you try with role field under properties  ? @roee88